File: scripts/reduce_chunk.py
import sys
import numpy as np
from cut_chunk_common import load_data
import os
import logging

def consolidate_remaps(remaps):
    nEntries = len(remaps)
    dic = dict()
    for i in range(nEntries):
        e = remaps[nEntries-i-1]
        dic[e['oid']] = dic.get(e['nid'], e['nid'])
    return dic

def load_remaps(fn):
    remaps = np.fromfile(fn, dtype=[('oid', np.uint64), ('nid', np.uint64)])
    return consolidate_remaps(remaps)

def load_sizes(ongoing, done):
    ongoing_counts = {x['sid']:x['s'] for x in np.fromfile(ongoing, dtype=[('sid', np.uint64), ('s', np.uint64)]) }
    done_counts = {x['sid']:x['s'] for x in np.fromfile(done, dtype=[('sid', np.uint64), ('s', np.uint64)]) }
    return ongoing_counts, done_counts

def minmax_remap_pair(pair, remaps):
    npair = [remaps.get(x, x) for x in pair]
    return (min(npair), max(npair))

def reduce_boundaries(tag, remaps, counts):
    reduced_map = dict()
    boundary_sv = set()
    for i in range(6):
        new_sids = []
        data = np.fromfile("boundary_{}_{}.data".format(i, tag), dtype=np.uint64)
        extra_data = np.fromfile("cut_plane_{}_{}.data".format(i, tag), dtype=np.uint64)
        logger.info("frozen sv before: {}".format(len(data)))
        nbs_seeds = set()
        for bs in data:
            boundary_sv.add(bs)
            nbs = remaps.get(bs, bs)
            if bs != nbs:
                reduced_map[bs] = nbs
            if nbs in counts:
                new_sids.append((bs, 1, nbs, counts[nbs]))
                nbs_seeds.add(nbs)
            else:
                logger.warning("face: {}; Impossible: no supervoxel count for {} ({})".format(i, nbs, bs))
        for bs in extra_data:
            nbs = remaps.get(bs, bs)
            if nbs in counts:
                nbs_seeds.add(nbs)
        for k in remaps:
            if remaps[k] in nbs_seeds and k not in reduced_map:
                new_sids.append((k, 0, remaps[k], counts[remaps[k]]))
                nbs_seeds.add(remaps[k])
        logger.info("frozen sv after: {}".format(len(new_sids)))
        np.array(new_sids,dtype=np.uint64).tofile("reduced_boundary_{}_{}.data".format(i, tag))

    return reduced_map, boundary_sv

def reduce_edges(tag, remaps):
    aff = load_data(os.environ['AFF_PATH'],mip=int(os.environ['AFF_MIP']))
    d_ie = [("s1", np.uint64), ("s2", np.uint64), ("aff", aff.dtype), ("area", np.uint64)]
    rg_array = np.fromfile("residual_rg_{}.data".format(tag), dtype=d_ie)
    edge_array = np.fromfile("incomplete_edges_{}.data".format(tag), dtype=d_ie)
    logger.info("Total number of edges: {}".format(len(rg_array)+len(edge_array)))
    merged_edges = dict()
    for l in [rg_array, edge_array]:
        for e in l:
            key = minmax_remap_pair((e['s1'], e['s2']), remaps)
            if key[0] == key[1]:
                continue;
            if key in merged_edges:
                merged_edges[key][0] += e['aff']
                merged_edges[key][1] += e['area']
            else:
                merged_edges[key] = [e['aff'], e['area']]

    logger.info("Total number of reduced edges: {}".format(len(merged_edges)))

    reduced_edges = []
    for k,v in merged_edges.items():
        logger.debug("new edge: {} {} {} {}".format(k[0],k[1],v[0],v[1]))
        reduced_edges.append((k[0], k[1], v[0], v[1]))

    np.array(reduced_edges, dtype=d_ie).tofile("reduced_edges_{}.data".format(tag))

def reduce_counts(tag, remaps, bs):
    dt = [('sid', np.uint64), ('s', np  .uint64)]
    sizes = np.fromfile("ongoing_supervoxel_counts_{}.data".format(tag), dtype=dt)
    reduced_sizes = dict()
    reduced_map = dict()
    for e in sizes:
        if e[0] in bs:
            continue
        sid = remaps.get(e[0], e[0])
        if sid != e[0]:
            reduced_map[e[0]] = sid
        if sid in reduced_sizes:
            reduced_sizes[sid] += e[1]
        else:
            reduced_sizes[sid] = e[1]

    np.array([(k,v) for k, v in reduced_sizes.items()], dtype=dt).tofile("reduced_ongoing_supervoxel_counts_{}.data".format(tag))

    return reduced_map


def reduce_sem(tag, remaps):
    nlabels = 3
    dt = [('sid', np.uint64), ('sem_labels', np.uint64, (nlabels,))]
    sems = np.fromfile("ongoing_semantic_labels_{}.data".format(tag), dtype=dt)
    for e in sems:
        e[0] = remaps.get(e[0], e[0])

    sems.tofile("reduced_ongoing_semantic_labels_{}.data".format(tag))


def reduce_size(tag, remaps):
    nlabels = 3
    dt = [('sid', np.uint64), ('size', np.uint64)]
    sizes = np.fromfile("ongoing_seg_size_{}.data".format(tag), dtype=dt)
    for e in sizes:
        e[0] = remaps.get(e[0], e[0])

    sizes.tofile("reduced_ongoing_seg_size_{}.data".format(tag))


def write_reduced_map(remaps):
    np.array([(k,v) for k, v in remaps.items()], dtype=[('os', np.uint64), ('ns', np.uint64)]).tofile("extra_remaps.data")


logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

remaps = load_remaps("remap.data")

tag = sys.argv[1]
ongoing_counts, done_counts = load_sizes("ongoing_segments.data", "done_segments.data")
counts = {**ongoing_counts, **done_counts}
reduced_map1, bs = reduce_boundaries(tag, remaps, counts)
reduced_map2 = reduce_counts(tag, remaps, bs)
reduce_edges(tag, remaps)
reduce_sem(tag, remaps)
reduce_size(tag, remaps)
write_reduced_map({**reduced_map1, **reduced_map2})
#print(load_segids("0_0_0_0"))
#bs_fname = sys.argv[1]
#bs_array = np.fromfile(bs_fname, dtype=[("chunkid", np.uint64), ("globalid", np.uint64)])

File: src/seg/reduce_chunk.cpp
#include <sys/stat.h>
#include <iostream>
#include <fstream>
#include <algorithm>
#include <cassert>
#include <execution>
#include <boost/format.hpp>
#include "Types.h"
#include "Utils.hpp"
#include "SemExtractor.hpp"

template <typename T, typename S>
struct __attribute__((packed)) sem_data_t
{
    T sid;
    S sem;
    sem_data_t() = default;
    explicit sem_data_t(const std::pair<T, S> & p) {
        sid = p.first;
        sem = p.second;
    }
};

template <typename T>
struct __attribute__((packed)) remap_data_t
{
    T oid;
    T nid;
    remap_data_t() = default;
    explicit remap_data_t(const std::pair<T, T> & p) {
        oid = p.first;
        nid = p.second;
    }
};

template <typename T>
struct __attribute__((packed)) size_data_t
{
    T sid;
    size_t size;
    size_data_t() = default;
    explicit size_data_t(const std::pair<T, size_t> & p) {
        sid = p.first;
        size = p.second;
    }
};

template <typename T>
void write_vector(const std::string & filename, const std::vector<T> & data)
{
    std::ofstream ofs(filename, std::ios_base::binary);
    for (auto const & e : data) {
        ofs.write(reinterpret_cast<const char *>(&(e)), sizeof(T));
    }
    assert(!ofs.bad());
    ofs.close();
}

template <typename T>
std::pair<T, T> minmax_remap_pair(const T & s1, const T & s2, const MapContainer<T, T> & remaps)
{
    auto nid1 = remaps.contains(s1) ? remaps.at(s1) : s1;
    auto nid2 = remaps.contains(s2) ? remaps.at(s2) : s2;
    return std::minmax(nid1, nid2);
}

template <typename T>
MapContainer<T, T> consolidate_remap(const std::vector<remap_data_t<T> > & remap_data)
{
    MapContainer<T, T> remaps;
    for (auto it = remap_data.rbegin(); it != remap_data.rend(); ++it) {
        remaps[it->oid] = remaps.contains(it->nid) ? remaps.at(it->nid) : it->nid;
    }
    return remaps;
}

template <typename T>
MapContainer<T, size_t> populate_size(const std::vector<size_data_t<T> > & ongoing_size_array, const std::vector<size_data_t<T> > & done_size_array)
{
    MapContainer<T, size_t> counts;
    for (auto & e : ongoing_size_array) {
        counts[e.sid] = e.size;
    }

    for (auto & e : done_size_array) {
        counts[e.sid] = e.size;
    }
    return counts;
}

template <typename T>
SetContainer<T> reduce_boundaries(const std::string & tag, const MapContainer<T, T> &remaps, const MapContainer<T, size_t> & counts, MapContainer<T, T> & reduced_map)
{
    SetContainer<T> boundary_sv;
    for (size_t i = 0; i < 6; i++) {
        std::vector<matching_entry_t<T> > new_sids;
        SetContainer<T> nbs_seeds;
        auto data = read_array<T>(str(boost::format("boundary_%1%_%2%.data") % i % tag));
        auto extra_data = read_array<T>(str(boost::format("cut_plane_%1%_%2%.data") % i % tag));
        std::cout << "frozen sv before: " << data.size() << std::endl;
        for (auto const & bs : data) {
            boundary_sv.insert(bs);
            auto nbs = remaps.contains(bs) ? remaps.at(bs): bs;
            if (bs != nbs) {
                reduced_map[bs] = nbs;
            }
            if (counts.contains(nbs)) {
                new_sids.push_back(matching_entry_t<T>{bs, 1, nbs, counts.at(nbs)});
                nbs_seeds.insert(nbs);
            }
        }
        for (auto const & bs : extra_data) {
            auto nbs = remaps.contains(bs) ? remaps.at(bs): bs;
            if (counts.contains(nbs)) {
                nbs_seeds.insert(nbs);
            }
        }
        if (not nbs_seeds.empty()) {
            for (auto const & [k, v] : remaps) {
                if (nbs_seeds.contains(v) and (!reduced_map.contains(k))) {
                    new_sids.push_back(matching_entry_t<T>{k, 0, v, counts.at(v)});
                }
            }
        }
        std::cout << "frozen sv after: " << new_sids.size() << std::endl;
        write_vector(str(boost::format("reduced_boundary_%1%_%2%.data") % i % tag), new_sids);
    }
    return boundary_sv;
}

template <typename T>
void reduce_counts(const std::string & tag, const MapContainer<T, T> & remaps, const SetContainer<T> & bs, MapContainer<T, T> & reduced_map)
{
    auto size_array = read_array<size_data_t<T> >(str(boost::format("ongoing_supervoxel_counts_%1%.data") % tag));
    MapContainer<T, size_t> reduced_size;
    for (auto const & e : size_array) {
        if (bs.contains(e.sid)) {
            continue;
        }
        auto sid = remaps.contains(e.sid) ? remaps.at(e.sid) : e.sid;
        if (sid != e.sid) {
            reduced_map[e.sid] = sid;
        }
        if (reduced_size.contains(sid)) {
            reduced_size[sid] += e.size;
        } else {
            reduced_size[sid] = e.size;
        }
    }
    write_vector(str(boost::format("reduced_ongoing_supervoxel_counts_%1%.data") % tag), std::vector<size_data_t<T> >(reduced_size.cbegin(), reduced_size.cend()));
}

void merge_edges(const auto & edges, const auto & remaps, auto & merged_edges)
{
    for (const auto & e : edges) {
        auto p = minmax_remap_pair(e.s1, e.s2, remaps);
        if (p.first == p.second) {
            continue;
        }
        if (merged_edges.contains(p)) {
            merged_edges[p].first += e.aff;
            merged_edges[p].second += e.area;
        } else {
            merged_edges[p] = std::make_pair(e.aff, e.area);
        }
    }
}

template <typename Ts, typename Ta>
void reduce_edges(const std::string & tag, const MapContainer<Ts, Ts> & remaps)
{
    auto rg_array = read_array<rg_entry<Ts, Ta> >(str(boost::format("residual_rg_%1%.data") % tag));
    auto edge_array = read_array<rg_entry<Ts, Ta> >(str(boost::format("incomplete_edges_%1%.data") % tag));
    std::cout << "Total number of edges: " << rg_array.size() + edge_array.size() << std::endl;
    MapContainer<SegPair<Ts>, std::pair<Ta, size_t>, HashFunction<SegPair<Ts> > > merged_edges;
    merge_edges(rg_array, remaps, merged_edges);
    merge_edges(edge_array, remaps, merged_edges);
    std::cout << "Total number of reduced edges: " << merged_edges.size() << std::endl;
    write_vector(str(boost::format("reduced_edges_%1%.data") % tag), std::vector<rg_entry<Ts, Ta> >(merged_edges.cbegin(), merged_edges.cend()));
}

template <typename T, typename S>
void reduce_sem(const std::string & tag, const MapContainer<T, T> & remaps)
{
    auto sem_array = read_array<sem_data_t<T, S> >(str(boost::format("ongoing_semantic_labels_%1%.data") % tag));
    for (auto & e : sem_array) {
        if (remaps.contains(e.sid)) {
            e.sid = remaps.at(e.sid);
        }
    }
    write_vector(str(boost::format("reduced_ongoing_semantic_labels_%1%.data") % tag), sem_array);
}

template <typename T>
void reduce_size(const std::string & tag, const MapContainer<T, T> & remaps)
{
    auto size_array = read_array<size_data_t<T> >(str(boost::format("ongoing_seg_size_%1%.data") % tag));
    for (auto & e : size_array) {
        if (remaps.contains(e.sid)) {
            e.sid = remaps.at(e.sid);
        }
    }
    write_vector(str(boost::format("reduced_ongoing_seg_size_%1%.data") % tag), size_array);
}

int main(int argc, char ** argv)
{
    std::string tag(argv[1]);
    auto remap_array = read_array<remap_data_t<seg_t> >("remap.data");
    auto remaps = consolidate_remap(remap_array);
    auto ongoing_size_array = read_array<size_data_t<seg_t> >("ongoing_segments.data");
    auto done_size_array = read_array<size_data_t<seg_t> >("done_segments.data");
    auto counts = populate_size(ongoing_size_array, done_size_array);
    MapContainer<seg_t, seg_t> reduced_map;
    auto boundary_sv =  reduce_boundaries(tag, remaps, counts, reduced_map);
    reduce_counts(tag, remaps, boundary_sv, reduced_map);
    reduce_edges<seg_t, aff_t>(tag, remaps);
    reduce_sem<seg_t, sem_array_t>(tag, remaps);
    reduce_size(tag, remaps);
    write_vector("extra_remaps.data", std::vector<remap_data_t<seg_t> >(reduced_map.cbegin(), reduced_map.cend()));
}
